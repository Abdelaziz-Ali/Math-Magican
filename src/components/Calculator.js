import React, { useState } from 'react';
import PropTypes, { string } from 'prop-types';
import calculate from '../logic/calculate';

function Calculator() {
  const [state, setState] = useState({ total: null, next: null, operation: null });

  const getBtn = (ButtonName) => {
    setState(calculate(state, ButtonName));
  };

  return (
    <div>
      <Display result={[state.total, state.next, state.operation]} />
      <Button getBtn={getBtn} />
    </div>
  );
}

function Display({ result }) {
  return (
    <h2 className="bg-gray color-white text-right pr-1 height-16" style={{ marginButton: '0px' }}>
      {result[0] == null && result[1] == null ? '0' : result[0]}
      {result[2] == null ? '' : result[2]}
      {result[1] == null ? '' : result[1]}
    </h2>
  );
}

Display.propTypes = {
  result: PropTypes.arrayOf(string).isRequired,
};

function Button({ getBtn }) {
  const buttons = [
    { id: 1, value: 'AC', color: 'gray-eo' },
    { id: 2, value: '+/-', color: 'gray-eo' },
    { id: 3, value: '%', color: 'gray-eo' },
    { id: 4, value: 'รท', color: 'orange' },
    { id: 5, value: '7', color: 'gray-eo' },
    { id: 6, value: '8', color: 'gray-eo' },
    { id: 7, value: '9', color: 'gray-eo' },
    { id: 8, value: 'x', color: 'orange' },
    { id: 9, value: '4', color: 'gray-eo' },
    { id: 10, value: '5', color: 'gray-eo' },
    { id: 11, value: '6', color: 'gray-eo' },
    { id: 12, value: '-', color: 'orange' },
    { id: 13, value: '1', color: 'gray-eo' },
    { id: 14, value: '2', color: 'gray-eo' },
    { id: 15, value: '3', color: 'gray-eo' },
    { id: 16, value: '+', color: 'orange' },
    { id: 17, value: '0', color: 'gray-eo col-span-2' },
    { id: 18, value: '.', color: 'gray-eo' },
    { id: 19, value: '=', color: 'orange' },
  ];

  const handelClick = (e) => {
    getBtn(e.target.value);
  };

  return (
    <div className="grid grid-cols-4 height-83">
      {
        buttons.map((button) => (
          <button value={button.value} onClick={handelClick} type="button" key={button.id} className={`bg-${button.color}`}>{button.value}</button>
        ))
      }
    </div>
  );
}

Button.propTypes = {
  getBtn: PropTypes.func.isRequired,
};

export default Calculator;
