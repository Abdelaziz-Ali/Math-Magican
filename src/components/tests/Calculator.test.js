import renderer from 'react-test-renderer';
import Calculator from '../Calculator';

it('renders Calculator component', () => {
  const component = renderer.create(<Calculator />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
