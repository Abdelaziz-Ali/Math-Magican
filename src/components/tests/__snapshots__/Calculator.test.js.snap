// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Calculator component 1`] = `
<div>
  <p
    className="bg-gray color-white text-right pr-1 height-16"
    style={
      Object {
        "fontFamily": "Seven Segment",
        "fontSize": "3em",
        "marginBlockEnd": "0em",
        "marginBlockStart": "0.4em",
      }
    }
  >
    0
  </p>
  <div
    className="grid grid-cols-4 height-83"
  >
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="AC"
    >
      AC
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="+/-"
    >
      +/-
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="%"
    >
      %
    </button>
    <button
      className="bg-orange"
      onClick={[Function]}
      type="button"
      value="รท"
    >
      รท
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="7"
    >
      7
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="8"
    >
      8
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="9"
    >
      9
    </button>
    <button
      className="bg-orange"
      onClick={[Function]}
      type="button"
      value="x"
    >
      x
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="4"
    >
      4
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="5"
    >
      5
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="6"
    >
      6
    </button>
    <button
      className="bg-orange"
      onClick={[Function]}
      type="button"
      value="-"
    >
      -
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="1"
    >
      1
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="2"
    >
      2
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="3"
    >
      3
    </button>
    <button
      className="bg-orange"
      onClick={[Function]}
      type="button"
      value="+"
    >
      +
    </button>
    <button
      className="bg-gray-eo col-span-2"
      onClick={[Function]}
      type="button"
      value="0"
    >
      0
    </button>
    <button
      className="bg-gray-eo"
      onClick={[Function]}
      type="button"
      value="."
    >
      .
    </button>
    <button
      className="bg-orange"
      onClick={[Function]}
      type="button"
      value="="
    >
      =
    </button>
  </div>
</div>
`;
